[PT2D][FSDP] skip FSDP hooks base on dynamo config

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122876
* __->__ #123021

unit test: `pytest test/distributed/_composable/fsdp/test_fully_shard_compile.py`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks`

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

weifengpy committed:
[FSDP2][PT2D] @torch.compiler.disable on FSDP hooks
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

yf225 commented on torch/distributed/_composable/fsdp/_fsdp_state.py:
```diff
@@ -142,6 +142,7 @@ def _init_fqns(self) -> None:
             if module in module_to_fsdp_param_group:
                 module_to_fsdp_param_group[module]._module_fqn = module_name
 
+    @torch.compiler.disable
```
The proposed changes will prevent `_pre_forward` and `_post_forward` from being traced during full-graph compile.

Discussed offline and a better solution is likely to disable only if `reshard_after_forward` is int.

awgu replied:
@yf225 I think one challenge here is that there are legitimate cases where the current graph-break tracing approach errors during tracing of the pre-forward hook, preventing using `torch.compile` at all.

Two cases I have seen for this are (1) DTensor-based 2D parallel and (2) most recently when I have been prototyping fp8 all-gather where the `Float8DynamicLinear.weight` is a separate tensor subclass.

I was thinking it may be worth considering some global config that allows us to switch between the graph-break and full-compile approaches since both may need to co-exist for some time (though let me know if you disagree).

yf225 replied:
Having a global config to switch between graph-break compile and full-graph compile sounds great. I can imagine some region doesn't work with graph-break compile but might be able to work with full-graph compile (and vice versa). We could use an `if` statement on the global config to decide whether to dynamo disable the hooks. Wonder wdyt?

awgu replied:
Global config and `if` statement sounds good to me!

weifengpy replied:
thanks for the discussion. I can add a dynamo config ``trace_fsdp_hooks=True/False``. An example UX would be 

```
with torch._dynamo.config.patch(trace_fsdp_hooks=False):
     model = torch.compile(model)
```

weifengpy committed:
Update on "[FSDP2][PT2D] @torch.compiler.disable on FSDP hooks"


for OSS FSDP2, users may try ``torch.compile(fsdp2_model)`` and we want it to work out of the box. Adding ``torch.compiler.disable`` to forward hooks for now

for pytorch internal, to compile fsdp hooks, maybe consdier removing ``torch.compiler.disable`` mannually

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2][PT2D] @torch.compiler.disable on FSDP hooks"


for OSS FSDP2, users may try ``torch.compile(fsdp2_model)`` and we want it to work out of the box. Adding ``torch.compiler.disable`` to forward hooks for now

for pytorch internal, to compile fsdp hooks, maybe consdier removing ``torch.compiler.disable`` mannually

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_compile.py:
```diff
@@ -0,0 +1,67 @@
+                self.assertTrue(hasattr(fsdp_hook, "_torchdynamo_disable"))
+                dynamo_ctx = inspect.getclosurevars(fsdp_hook).nonlocals["self"]
+                self.assertTrue(
+                    isinstance(dynamo_ctx, torch._dynamo.eval_frame.DisableContext)
```
the function should be wrapped by ``DisableContext`` with ``callback=None``

weifengpy committed:
Update on "[FSDP2][PT2D] skip compiling FSDP hooks by dynamo config"


This PR enables torch.compile to skip or trace into FSDP hooks by config
```
torch._dynamo.config.skip_fsdp_hooks = True
torch.compile(model) 
# OR
# model.compile()
``` 

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2][PT2D] _dynamo.config.skip_fsdp_hooks for torch.compile and nn.Module.compile"


This PR enables torch.compile to skip or trace into FSDP hooks by config. This should unblock FSDP2 CI (eager/skip hooks/full compile), and fp8 (FSDP extention in general). It helps debugging as well by comparing dynamo behavior between  ``skip_fsdp_hooks = True/False``
 
```
torch._dynamo.config.skip_fsdp_hooks = True
torch.compile(model) 
# OR
# model.compile()
``` 


**what about setting torch.dynamo.disable in FSDP2**: ``torch.dynamo.disable`` is applied at import time statically. We want a flexiable way to skip or trace into FSDP2 hooks, 

**what about existing ruls in trace_rules.py**: FSDP hooks and TP/SP hooks all share DTensor code path. We want to skip FSDP hooks without skipping TP/SP hooks. Thus did not choose file or torch obj based rules

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2][PT2D] _dynamo.config.skip_fsdp_hooks for torch.compile and nn.Module.compile"


This PR enables torch.compile to skip or trace into FSDP hooks by config. This should unblock FSDP2 CI (eager/skip hooks/full compile), and fp8 (FSDP extention in general). It helps debugging as well by comparing dynamo behavior between  ``skip_fsdp_hooks = True/False``
 
```
torch._dynamo.config.skip_fsdp_hooks = True
torch.compile(model) 
# OR
# model.compile()
``` 


**what about setting torch.dynamo.disable in FSDP2**: ``torch.dynamo.disable`` is applied at import time statically. We want a flexiable way to skip or trace into FSDP2 hooks, 

**what about existing ruls in trace_rules.py**: FSDP hooks and TP/SP hooks all share DTensor code path. We want to skip FSDP hooks without skipping TP/SP hooks. Thus did not choose file or torch obj based rules

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2][PT2D] _dynamo.config.skip_fsdp_hooks for torch.compile and nn.Module.compile"


This PR enables torch.compile to skip or trace into FSDP hooks by config. This should unblock FSDP2 CI (eager/skip hooks/full compile), and fp8 (FSDP extention in general). It helps debugging as well by comparing dynamo behavior between  ``skip_fsdp_hooks = True/False``
 
```
torch._dynamo.config.skip_fsdp_hooks = True
torch.compile(model) 
# OR
# model.compile()
``` 


**what about setting torch.dynamo.disable in FSDP2**: ``torch.dynamo.disable`` is applied at import time statically. We want a flexiable way to skip or trace into FSDP2 hooks, 

**what about existing ruls in trace_rules.py**: FSDP hooks and TP/SP hooks all share DTensor code path. We want to skip FSDP hooks without skipping TP/SP hooks. Thus did not choose file or torch obj based rules

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2][PT2D] _dynamo.config.skip_fsdp_hooks for torch.compile and nn.Module.compile"


This PR enables torch.compile to skip or trace into FSDP hooks by config. This should unblock FSDP2 CI (eager/skip hooks/full compile), and fp8 (FSDP extention in general). It helps debugging as well by comparing dynamo behavior between  ``skip_fsdp_hooks = True/False``
 
```
torch._dynamo.config.skip_fsdp_hooks = True
torch.compile(model) 
# OR
# model.compile()
``` 


**what about setting torch.dynamo.disable in FSDP2**: ``torch.dynamo.disable`` is applied at import time statically. We want a flexiable way to skip or trace into FSDP2 hooks, 

**what about existing ruls in trace_rules.py**: FSDP hooks and TP/SP hooks all share DTensor code path. We want to skip FSDP hooks without skipping TP/SP hooks. Thus did not choose file or torch obj based rules

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy committed:
Update on "[Dynamo] @torch.compiler.disable_if_config_true to skip functions"


`torch.compiler.disable` is executed when defining functions. To turn on/off base on `torch._dynamo.config` during CI, implementing function wrapper `torch.compiler.disable_if_config_true`

```
torch.compiler.disable_if_config_true("your_attr_name_from_dynamo_config")
def func(x):
    return x.sin().cos()
```

unit test: `pytest test/dynamo/test_decorators.py -k test_disable_if_config_true`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks` to test againt compiling compute and full tracing. By default we trace into hooks with `skip_fsdp_hooks=False`
```
torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _pre_forward

torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _post_forward
```

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

yf225 commented on torch/_dynamo/config.py:
```diff
@@ -289,6 +289,7 @@ def _get_optimize_ddp_mode():
 
 # Whether to skip guarding on FSDP-managed modules
 skip_fsdp_guards = True
+skip_fsdp_hooks = False
```
would be great to have comment on this describing what this config controls

weifengpy replied:
added comment

yf225 commented on torch/_dynamo/decorators.py:
```diff
@@ -67,6 +68,20 @@ def skip(fn=None):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if getattr(config, config_attr, None):
+                return disable(func)(*args, **kwargs)
```
Do we need `disable(recursive=True)` here?

weifengpy replied:
it's calling `disable` and `disable` has `recursive=True`. but I can be explicit here with `recursive=True`

weifengpy committed:
Update on "[Dynamo] @torch.compiler.disable_if_config_true to skip functions"


`torch.compiler.disable` is executed when defining functions. To turn on/off base on `torch._dynamo.config` during CI, implementing function wrapper `torch.compiler.disable_if_config_true`

```
torch.compiler.disable_if_config_true("your_attr_name_from_dynamo_config")
def func(x):
    return x.sin().cos()
```

unit test: `pytest test/dynamo/test_decorators.py -k test_disable_if_config_true`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks` to test againt compiling compute and full tracing. By default we trace into hooks with `skip_fsdp_hooks=False`
```
torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _pre_forward

torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _post_forward
```

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

anijain2305 commented on torch/compiler/__init__.py:
```diff
@@ -97,6 +98,17 @@ def disable(fn=None, recursive=True):
 
     return torch._dynamo.disable(fn, recursive)
 
+def disable_if_config_true(config_attr: str):
```
This API should be private. Its current usage is limited to FSDP hooks. And its utility as for an torch.compile user is not fully proven yet. We should have something similar to `torch._dynamo.disable`.

anijain2305 commented on torch/distributed/_composable/fsdp/_fsdp_state.py:
```diff
@@ -161,6 +162,7 @@ def _pre_forward(
             args, kwargs = self._fsdp_param_group.pre_forward(module, args, kwargs)
         return args, kwargs
 
+    @torch.compiler.disable_if_config_true("skip_fsdp_hooks")
```
Do you plan to use this API for anything apart from `skip_fsdp_hooks`? If not, I would suggest to just define the decorator at the top of this file. And just keep it really simple

```
def disable_if_config_true(func):
        @functools.wraps(func)
        def inner(*args, **kwargs):
            if config.skip_fsdp_hooks:
                return disable(func)(*args, **kwargs)
            else:
                return func(*args, **kwargs)

        return inner
```

anijain2305 reviewed (approved):
As of now, the API seems very specific to `skip_fsdp_hooks`. So, I suggest that we don't expose this API anywhere else. And just put all the code in `torch/distributed/_composable/fsdp/_fsdp_state.py`

anijain2305 reviewed (changes requested):
As of now, the API seems very specific to `skip_fsdp_hooks`. So, I suggest that we don't expose this API anywhere else. And just put all the code in `torch/distributed/_composable/fsdp/_fsdp_state.py`

yf225 commented:
From PT2-D meeting: do we need a similar approach to disable hooks when an nn.Module is disabled? we likely need to figure out the answer to this. cc. @weifengpy @anijain2305

weifengpy committed:
Update on "[Dynamo] @torch.compiler.disable_if_config_true to skip functions"


`torch.compiler.disable` is executed when defining functions. To turn on/off base on `torch._dynamo.config` during CI, implementing function wrapper `torch.compiler.disable_if_config_true`

```
torch.compiler.disable_if_config_true("your_attr_name_from_dynamo_config")
def func(x):
    return x.sin().cos()
```

unit test: `pytest test/dynamo/test_decorators.py -k test_disable_if_config_true`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks` to test againt compiling compute and full tracing. By default we trace into hooks with `skip_fsdp_hooks=False`
```
torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _pre_forward

torch.compiler.disable_if_config_true("skip_fsdp_hooks")
def _post_forward
```

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy commented:
> From PT2-D meeting: do we need a similar approach to disable hooks when an nn.Module is disabled? we likely need to figure out the answer to this. cc. @weifengpy @anijain2305

agree. my understanding is, if we apply `torch._dynamo.disable(nn.Module)`
* forward hooks should be disable without the need to explicit disable hooks. Otherwise it's a bug (very unlikely) @anijain2305 
* backword hooks should be in eager mode (no eval fram) unless they are using compiled autograd

weifengpy committed:
Update on "[PT2D][FSDP] skip FSDP hooks base on dynamo config"


unit test: `pytest test/distributed/_composable/fsdp/test_fully_shard_compile.py`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_compile.py:
```diff
@@ -0,0 +1,60 @@
+            self.assertEqual(trace_rules_check_count, 0)
+        else:
+            self.assertTrue(trace_rules_check_count > 0)
+
```
FSDP hooks are in skipfiles already. we do counting in trace rules instead of using ``torch._dynamo.testing.CompileCounter``

anijain2305 commented on torch/distributed/_composable/fsdp/_fsdp_state.py:
```diff
@@ -36,6 +36,17 @@ def __init__(self):
 
 
+def disable_if_config_true(func):
+    @functools.wraps(func)
+    def fsdp_hook_wrapper(*args, **kwargs):
+        if torch._dynamo.config.skip_fsdp_hooks:
+            return torch._dynamo.disable(func, recursive=True)(*args, **kwargs)
+        else:
+            return func(*args, **kwargs)
+
+    return fsdp_hook_wrapper
```
I think this can be simplified as 

```
if config...:
	return torch._dynamo.disable(func)
else:
        return func
```

No need of inner wrapper.

weifengpy replied:
we discussed about this: we are using ``@functools.wraps(func)`` and ``torch._dynamo.disable(func, recursive=True)(*args, **kwargs)`` because we want a chance to set `_dynamo.config` before calling the function.

weifengpy commented:
> From PT2-D meeting: do we need a similar approach to disable hooks when an nn.Module is disabled? we likely need to figure out the answer to this. cc. @weifengpy @anijain2305

@yf225  @anijain2305  actually ``torch.compiler.disable(nn.Module)`` indeed cannot disable hooks. filed a minimal repro https://github.com/pytorch/pytorch/issues/123979

weifengpy committed:
Update on "[PT2D][FSDP] skip FSDP hooks base on dynamo config"


unit test: `pytest test/distributed/_composable/fsdp/test_fully_shard_compile.py`

For FSDP, we turn on/off compiling hooks base on `torch._dynamo.config.skip_fsdp_hooks`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge
